DEVX-1034 Adds GKE/Kubernetes Confluent Operator base demonstration (#257)
confluentinc/examples: kubernetes/gke-base/docs/index.rst

.. _quickstart-demos-operator-gke

|cp| on Google Kubernetes Engine
======================================

Overview
--------

> This demo shows a deployment of |cp| on Google Kubernetes Engine (GKE) leveraging |co-long| with random data generation provided via the `Kafka Connect Datagen <https://www.confluent.io/hub/confluentinc/kafka-connect-datagen>`__.

Reviewer: For consideration: change all mentions of `random` to `mock`

+------------------+----------------+---------------------------------------------------------+
| ``jq``           | ``1.6``        | https://stedolan.github.io/jq/                          |
+------------------+----------------+---------------------------------------------------------+
| ``gcloud``       | ``259.0.0``    |  https://cloud.google.com/sdk/install                   |
| ``GCP sdk core`` | ``2019.08.23`` |                                                         |
+------------------+----------------+---------------------------------------------------------+

Running the Demo
----------------

.. warning:: This demo uses the real GCP SDK to launch real resources. To avoid unexpected charges, carefully evaluate the cost of resources before launching the demo and ensure all resources are destroyed after you are done evaluating the demonstration.  Optionally, refer to the |co| :ref:`Sizing Recommendations <co-env-sizing>` document and the :ref:`examples-operator-gke-base-variable-reference` section for more information on required resources for running |cp| on Kubernetes.

 
Setup
*****

> Clone the Confluent examples repository and change directories on your termianl into the ``gke-base`` directory.

Reviewer: ```suggestion
Clone the Confluent examples repository and change directories on your terminal into the ``gke-base`` directory.
```

.. sourcecode:: bash

> 		git clone git@github.com:confluentinc/examples.git; cd examples/operator/gke-base

Reviewer: ```suggestion
		git clone git@github.com:confluentinc/examples.git
		cd examples/operator/gke-base
```

Author: I kind of like it in one line for copy and paste purposes... thoughts?

Reviewer: 1. Docs now have a copy/paste button for easy copy.

2. Generally speaking, examples should err on the side of being more explicit to help users learn.

Author: I thought users might run into issues if multiple commands are in the same "code snippet" block, which is why I did this one this way.  I'm not certain any/all commands will have issues, so I don't have evidence this for sure is an issue.  I've definitely tried to keep command output out of the snippet blocks for this reason.   If you think it's safe to "stack" commands like this I'll break it up

Reviewer: Yeah, "stacking" commands is easier to grok, especially for new users.  But this is just a nit, do what you deem better.

Author: I tested them stacked and didn't see any issue.  If we see field issues I can adjust accordingly.

    ...
    Created [https://container.googleapis.com/v1/projects/<project-id>/zones/us-central1-a/clusters/cp-examples-operator-<username>].
    To inspect the contents of your cluster, go to: <link> 
    kubeconfig entry generated for cp-examples-operator-<username>.
    NAME                            LOCATION  MASTER_VERSION  MASTER_IP     MACHINE_TYPE  NODE_VERSION   NUM_NODES  STATUS
    cp-examples-operator-<username> <zone>    1.12.8-gke.10   <ip-address>  n1-highmem-2  1.12.8-gke.10  3          RUNNING
    ✔  ++++++++++ GKE Cluster Created

.. _examples-operator-gke-base-validate:

Validate
********

The demo uses ``kubectl`` to control the cluster.  To verify that your local ``kubectl`` configured as intended, run:

.. sourcecode:: bash

    kubectl config current-context

The context should contain the proper region and cluster name.  If you used the demo ``gke-create-cluster`` function to create your cluster, the context name should have the format: ``gke_<google-project-id>_<region>_<cp-examples-operator>-<username>``

.. _examples-operator-gke-base-run:

Run
***

To deploy |cp| run (estimated running time, 7 minutes):

.. sourcecode:: bash

    make demo

> The last output message you should see should be::

Reviewer: ```suggestion
The last output message you should see is:
```

You could view the output of the random click data generator with the console consumer::

	kafka-console-consumer --bootstrap-server kafka:9071 --consumer.config /etc/kafka-client-properties/kafka-client.properties --topic clicks

Example output might look like::

	222.152.45.45F-
	16141<GET /images/track.png HTTP/1.1204006-Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/59.0.3071.115 Safari/537.36
	122.173.165.203L-
	16151FGET /site/user_status.html HTTP/1.1401289-Mozilla/5.0 (compatible; Googlebot/2.1; +http://www.google.com/bot.html)
	...

Verify Confluent Platform Control Center
````````````````````````````````````````

In order to view |c3|, network connectivity will need to be available between your local machine and the Kubernetes pod running the |c3| service.  If you used an existing cluster you may already have external cluster access configured.  If you used the demo ``gke-create-cluster`` function, you can use the following ``kubectl`` command to open a forwarded port connection between your local host and |c3|.

.. sourcecode:: bash

		kubectl -n operator port-forward controlcenter-0 12345:9021

> Now open a web-browser to http://localhost:12345, and you should see |c3| with operational |ak| clusters, |sr|, and |kconnect-long|.

Reviewer: ```suggestion
Now open a web-browser to http://localhost:12345, and you should see |c3| with your operational |ak| cluster, |sr|, and |kconnect-long|.
```

> .. figure:: images/c3.png

Reviewer: 1. Can we get the topic **inspection** view, instead of the topic overview?
2. For consideration, add a view that shows the topic schema (nudge folks to use Schema Registry)

Author: @ybyzek updated w/ new screenshots, let me know what you think of those
